src/compiler/sourcemap.ts

```diff
@@ -231,34 +232,37 @@ namespace ts {
                 Debug.assertEqual(lastGeneratedLine, pendingGeneratedLine, "generatedLine cannot backtrack");
                 // Emit comma to separate the entry
                 if (hasLast) {
-                    mappings += ",";
+                    mappingSegmentsBuffer.push(",");
                 }
             }
 
             // 1. Relative generated character
-            mappings += base64VLQFormatEncode(pendingGeneratedCharacter - lastGeneratedCharacter);
+            mappingSegmentsBuffer.push(base64VLQFormatEncode(pendingGeneratedCharacter - lastGeneratedCharacter));
             lastGeneratedCharacter = pendingGeneratedCharacter;
 
             if (hasPendingSource) {
                 // 2. Relative sourceIndex
-                mappings += base64VLQFormatEncode(pendingSourceIndex - lastSourceIndex);
+                mappingSegmentsBuffer.push(base64VLQFormatEncode(pendingSourceIndex - lastSourceIndex));
                 lastSourceIndex = pendingSourceIndex;
 
                 // 3. Relative source line
-                mappings += base64VLQFormatEncode(pendingSourceLine - lastSourceLine);
+                mappingSegmentsBuffer.push(base64VLQFormatEncode(pendingSourceLine - lastSourceLine));
                 lastSourceLine = pendingSourceLine;
 
                 // 4. Relative source character
-                mappings += base64VLQFormatEncode(pendingSourceCharacter - lastSourceCharacter);
+                mappingSegmentsBuffer.push(base64VLQFormatEncode(pendingSourceCharacter - lastSourceCharacter));
                 lastSourceCharacter = pendingSourceCharacter;
 
                 if (hasPendingName) {
                     // 5. Relative nameIndex
-                    mappings += base64VLQFormatEncode(pendingNameIndex - lastNameIndex);
+                    mappingSegmentsBuffer.push(base64VLQFormatEncode(pendingNameIndex - lastNameIndex));
                     lastNameIndex = pendingNameIndex;
                 }
             }
 
+            mappings = mappingSegmentsBuffer.join("");

```

Comment:
In local experimentation, I've found that I can get about a 40% speed up by batching the char code conversion:
```
function base64VLQFormatEncode2(value) {
  if (value < 0) {
      value = ((-value) << 1) + 1;
  }
  else {
      value = value << 1;
  }

  // Encode 5 bits at a time starting from least significant bits
  const result = [];
  do {
      let currentDigit = value & 31; // 11111
      value = value >> 5;
      if (value > 0) {
          // There are still more digits to decode, set the msb (6th bit)
          currentDigit = currentDigit | 32;
      }
      result.push(base64FormatEncode(currentDigit));
  } while (value > 0);

  return result;
}

function buildMappingStr() {
  const charCodes = [];
  // Mock representation of adding a segment, for benchmarking
  for (let i = 0; i < target; i ++) {
    charCodes.push(...base64VLQFormatEncode2(i));
  }
  let str = '';
  const segmentLength = 4096;
  for (let i = 0, len = charCodes.length; i < len; i += segmentLength) {
    const segment = String.fromCharCode.apply(null, charCodes.slice(i, i + segmentLength));
    str += segment;
  }
  return str;
}
```

Reply:
If you are encoding a very large number of mappings (>10000), it gets faster to use an encoder with its own buffer:
```
class SourceMapEncoder {
  constructor() {
    this._encoded = '';
    this._buffer = new Uint8Array(1024);
    this._bufferPos = 0;
  }

  get encoded() {
    if (this._bufferPos > 0) {
      this._commit(); 
    }
    return this._encoded;
  }

  /**
   * @param {number} code
   */
  appendCharCode(code) {
    this._buffer[this._bufferPos] = code;
    this._bufferPos++;
    
    if (this._bufferPos === this._buffer.length) {
      this._commit();
    }
  }

  /**
   * @param {number} value
   */
  appendBase64VLQ(value) {
    if (value < 0) {
        value = ((-value) << 1) + 1;
    }
    else {
        value = value << 1;
    }
  
    // Encode 5 bits at a time starting from least significant bits
    do {
        let currentDigit = value & 31; // 11111
        value = value >> 5;
        if (value > 0) {
            // There are still more digits to decode, set the msb (6th bit)
            currentDigit = currentDigit | 32;
        }
        this.appendCharCode(formatArr[currentDigit]);
    } while (value > 0);
  }

  reset() {
    this._encoded = '';
    this._bufferPos = 0;
  }

  _commit() {
    this._encoded += String.fromCharCode.apply(null, this._buffer.slice(0, this._bufferPos));
    this._bufferPos = 0;
  }
}
```
Call `.appendBase64VLQ(value)` for each mapping to append, then retrieve `.encoded` at the end.
Still trying to figure out why the overhead is so much higher for small counts

Reply:
That's pretty funny, that's what I've been working on separately over the last hour - check out https://github.com/microsoft/TypeScript/pull/43987